Remove i18n-calypso moment from state.

sgomes opened the pull request:
This PR is one of many slowly working towards removing the `moment` export from `i18n-calypso`.

It switches state to native dates, vanilla `moment`, or vanilla `moment` with applied locale, depending on what's possible in each case.

Note that these selectors can potentially return incorrectly localized results because of caching (this was already the case, and this PR doesn't change that). It would be good to explore moving localization outside of selectors to avoid that situation, or, alternatively, to find a way of invalidating these selectors' caches when the locale changes. Either option would be out of scope for this PR.

Please feel free to add whomever you feel should be included in the reviewers list.

#### Changes proposed in this Pull Request

* Replace `i18n-calypso` `moment` with native dates, vanilla `moment`, or vanilla `moment` with applied locale, depending on what's possible in each case.

#### Testing instructions

I am not familiar with most of this functionality, so I am unable to provide useful testing steps. I'm hoping one of the reviewers will know more. State is usually covered by unit tests, but not extensively, in this case.

The main things to ensure are that the modified code doesn't produce any errors, and that any time-related strings are still correctly localized.

sgomes committed:
Remove i18n-calypso moment from state.

Switch to native dates, vanilla moment, or vanilla moment with applied
locale, depending on what's possible in each case.

sgomes commented on client/state/selectors/get-site-stats-view-summary.js:
```diff
@@ -1,11 +1,8 @@
-
-import { moment } from 'i18n-calypso';
 import { round } from 'lodash';
+import moment from 'moment';
```
This file shouldn't require any date localization.

sgomes commented on client/state/stats/lists/selectors.js:
```diff
@@ -1,11 +1,8 @@
-
 import { get, isArray, map, flatten } from 'lodash';
-import i18n from 'i18n-calypso';
+import moment from 'moment';
```
This file shouldn't require any date localization.

sgomes commented:
@jsnajdr Given your review comments in #37073, I'd like to improve things in this one regarding currently incorrect code.

Some of the selectors being rewritten here return localised date strings, which can potentially become stale as the user changes locales.

The easier approach to solve this is to invalidate the cache for these selectors, e.g. by making use of the `getCacheKey` parameter in `createSelector`. Do you think this is a reasonable approach, or should the selectors be rewritten to not return localised date strings?

sgomes commented:
@jsnajdr Still waiting on feedback on my comment above. Any chance you could weigh in, so I could get started on it?

blowery commented:
> The easier approach to solve this is to invalidate the cache for these selectors, e.g. by making use of the getCacheKey parameter in createSelector. Do you think this is a reasonable approach, or should the selectors be rewritten to not return localised date strings at all?

I think, ideally, selectors don't return localized strings. That's up to the view to do, which has reactive access to the current locale. 

That said, hacking in support using the cache key is probably mostly ok. I just worry that other folks will see it and use it as a template for future work.

How much of a diff is there in work between the two solutions?

sgomes commented:
> How much of a diff is there in work between the two solutions?

It highly depends on how many components are making use of said selectors. I'll need to look into that.

In any case, I do agree with the concerns you've pointed out, so I'll see if I can go about it the "right" way! 👍

sgomes committed:
getBillingTransactionsByType: no localized moment.

sgomes committed:
Locale changes on getFilteredBillingTransactions.

sgomes committed:
Fix new lint issue.

sgomes committed:
getScheduledPublicizeShareActionTime: no moment.

sgomes committed:
Remove unnecessary mock.

sgomes committed:
enrichPublicizeActionsWithConnections: no moment.

sgomes commented:
Rebased on master after `@format` pragma codemod.

sgomes commented:
I think I've pretty much taken care of what I could here. Ready for review, @jsnajdr and @blowery!

sgomes committed:
Fix a whole bunch of lint issues.

sgomes commented:
Rebased on master.

jsnajdr commented on client/blocks/post-share/index.jsx:
```diff
@@ -277,7 +278,7 @@ class PostShare extends Component {
 				type: 'published-action',
 				title: message,
-				date,
+				date: moment( date ),
```
In the objects passed as items in `events`, the `date` field is expected to be `moment` or `Date` instance or a number timestamp. See here: https://github.com/Automattic/wp-calypso/blob/master/client/components/date-picker/index.jsx#L182-L192

That means we don't need `moment` here and `new Date( date )` is just fine. What is the `date` destructured from input in this `map` statement? A string? We could also patch `DatePicker.getDateInstance` to accept a string, too.

sgomes replied:
I've been unsuccessfully trying to test this for about half an hour no, and am no closer to being able to tell you what `date` is here. This is a good example of where types would be useful 😞

Is there a good way of testing this without creating and linking a social media account?

sgomes replied:
From what I can tell, the original code was wrong, and this shouldn't be `date`, but rather `share_date` or `shareDate`, which according to the JSON schema is an integer (so UNIX timestamp, perhaps?).

sgomes replied:
@jsnajdr: Let me know if you find anything out here, happy to make changes. Otherwise, this is one place I'd suggest keeping the current high-level approach of trying to preserve the current behaviour, rather than going too in-depth.

jsnajdr replied:
@sgomes Would you mind if I push a few commits if I find anything during testing? I have a few Facebook testing pages connected to my account.

sgomes replied:
@jsnajdr: No, please go right ahead! And thank you! 🙇 

I'll work on a separate PR in the meantime, tracking down the a few more bits of `i18n-calypso` `moment` in state that I found.

jsnajdr committed:
PostShare: return raw timestamp from the selector that returns status of share action

After we just performed a share action, the result (success/failure, share date) is stored
in Redux and used by `PostShare` to display a notice.

Change the selector to return the raw Unix timestamp and convert it to `Date` or `moment` only
in the React UI component. That makes sure that the selector returns identical value on each call
and doesn't cause rerenders.

jsnajdr commented on client/state/selectors/utils/index.js:
```diff
@@ -15,26 +13,25 @@ import getPublicizeConnection from 'state/selectors/get-publicize-connection';
 				result,
-				service: get( connection, 'service', 'twitter' ),
-				shareDate: actionDate.format( 'llll' ),
+				service: connection?.service?.twitter,
```
The correct expression here is `connection?.service ?? 'twitter'`. I'm fixing it locally and will push a commit. Discovered this bug when searching where have all the social icons gone 🙂

sgomes replied:
Ahhh, good catch! I missed that it was a third parameter and not a string array 🤦‍♂

jsnajdr committed:
enrichPublicizeActionsWithConnections: fix social icon, don't parse date

Fixes code that gets the social icon slug and avoid parsing the date string
to a `Date` object. All usages of the `date` field pass it through the `moment`
constructor anyway.

jsnajdr committed:
DatePicker: simplify the getDateInstance implementation

Instead of detecting all possible date formats, just delegate that to moment.js
by passing the date value through `moment` constructor.

jsnajdr committed:
PostShare: don't parse share date with moment.js

All usages will pass the value through the `moment` constructor anyway.

jsnajdr committed:
Billing History: fix rendering of selected filter title

The title rendering got changed and was correctly patched only for the select dropdown items,
but not for the selected state when the dropdown is closed.

jsnajdr committed:
getBillingTransactions: don't parse the transaction date string

All usages of the `date` string, both displaying and using in filters, goes
through the `moment()` constructor.

jsnajdr committed:
Billing Transactions: fixup searching by human date, add unit test

jsnajdr commented on client/state/stats/lists/selectors.js:
```diff
@@ -77,7 +76,7 @@ export const getSiteStatsPostStreakData = treeSelect(
 		if ( streakData && streakData.data && ! isArray( streakData.data ) ) {
 			Object.keys( streakData.data ).forEach( timestamp => {
-				const postDay = i18n.moment.unix( timestamp ).locale( 'en' );
+				const postDay = moment.unix( timestamp ).locale( 'en' );
```
We don't need the `locale( 'en' )` call here. The format string `YYYY-MM-DD` doesn't include any i18n-dependent info -- just universal numbers.

sgomes replied:
SGTM! 👍

jsnajdr committed:
getSiteStatsPostStreakData: remove unneeded locale('en') call

jsnajdr commented:
I tested all the changes in this PR and pushed 8 more commits with fixes 🙂 

In the `PostShare` part, I removed date parsing (i.e., creating `Date` and `moment` instances) from selectors, in favor of returning raw data. And verified that all usages are OK with that, as they always pass the value through a `moment` constructor anyway.

That helps with avoiding unneeded rerenders, because `new Date` or `moment()` return a new object instance on each invocation, even if the input is identical.

Then I fixed a little bug where social icons in the sharing tooltip were broken -- bad conversion of `get()` call to optional chaining.

In the Billing Transaction section, I fixed a bug where a change in how select dropdown items are rendered broke the display of closed dropdown:

<img width="432" alt="Screenshot 2019-12-17 at 10 03 00" src="https://user-images.githubusercontent.com/664258/70985921-3f3b9a80-20bd-11ea-9536-b0e8acb49b6d.png">

I also removed date parsing from the `getBillingTransactions` selector, making the selector trivial. Its unit tests were no longer testing anything interesting, so I removed them.

The date parsing removal required a fixup of the code that filters billing transaction by formatted date like 'May 1'. I added a unit test for that -- there wasn't one before.

I think this PR is now ready to ship. @sgomes what do you think?

jsnajdr reviewed (approved):
:+1:

sgomes commented:
This all looks great to me! Those were some excellent cleanups, @jsnajdr! 👍 

Thank you for all your help! 🙇 

We're nearly there with the `i18n-calypso` `moment` stuff. I'm working on one more PR with state stuff, which I'll be publishing in a while.
